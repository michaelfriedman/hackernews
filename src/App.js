import React, { Component } from 'react'
import Search from './components/Search'
import Table from './components/Table'
import Button from './components/Button'
import Loading from './components/Loading'
import ButtonWithLoading from './components/ButtonWithLoading'
import './App.css'

const DEFAULT_QUERY = 'redux'
const DEFAULT_PAGE = 0
const DEFAULT_HPP = '100'
const PATH_BASE = 'https://hn.algolia.com/api/v1'
const PATH_SEARCH = '/search'
const PARAM_SEARCH = 'query='
const PARAM_PAGE = 'page='
const PARAM_HPP = 'hitsPerPage='

class App extends Component {
  constructor (props) {
    super(props)

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      isLoading: false,
      sortKey: 'NONE',
      isSortReverse: false
    }
    this.onDismiss = this.onDismiss.bind(this)
    this.onSearchChange = this.onSearchChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this)
    this.onSort = this.onSort.bind(this)
  }

  componentDidMount () {
    const { searchTerm } = this.state
    this.setState({ searchKey: searchTerm })
    this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE)
  }

  fetchSearchTopStories (searchTerm, page) {
    this.setState({ isLoading: true })
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
  }

  needsToSearchTopStories (searchTerm) {
    return !this.state.results[searchTerm]
  }

  onSearchSubmit (event) {
    event.preventDefault()
    const { searchTerm } = this.state
    this.setState({ searchKey: searchTerm })
    if (this.needsToSearchTopStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE)
    }
  }

  setSearchTopStories (result) {
    const { hits, page } = result
    const { searchKey, results } = this.state

    const oldHits = results && results[searchKey]
      ? results[searchKey].hits
      : []

    const updatedHits = [...oldHits, ...hits]

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      },
      isLoading: false
    })
  }

  onSearchChange (event) {
    this.setState({
      searchTerm: event.target.value
    })
  }

  onDismiss (id) {
    const { searchKey, results } = this.state
    const { hits, page } = results[searchKey]
    const isNotId = item => item.objectID !== id
    const updatedHits = hits.filter(isNotId)
    this.setState({
      results: {
        ...results,
        [searchKey]: {
          hits: updatedHits,
          page
        }
      }
    })
  }

  onSort (sortKey) {
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse
    this.setState({ sortKey, isSortReverse })
  }

  render () {
    const {
      searchTerm,
      results,
      searchKey,
      isLoading,
      isSortReverse,
      sortKey
    } = this.state
    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0
    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || []

    return (
      <div className='page'>
        <div className='interactions'>
          <Search
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
            value={searchTerm}>
            Search
          </Search>
        </div>
        <Table
          isSortReverse={isSortReverse}
          list={list}
          onDismiss={this.onDismiss}
          onSort={this.onSort}
          sortKey={sortKey}
        />
        <div className='interactions'>
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
            More
          </ButtonWithLoading>
        </div>
      </div>
    )
  }
}

export default App

export {
  Button,
  Search,
  Table,
  Loading,
  ButtonWithLoading
}
