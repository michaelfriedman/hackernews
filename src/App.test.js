import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import App, { Search, Button, Table, Loading, ButtonWithLoading } from './App'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
})

test('snapshots', () => {
  const component = renderer.create(
    <App />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('Search', () => {
  it('renders', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Search>Search</Search>, div)
  })
  test('snapshots', () => {
    const component = renderer.create(
      <Search>Search</Search>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Button', () => {
  it('renders', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Button>More</Button>, div)
  })
  test('snapshots', () => {
    const component = renderer.create(
      <Button>More</Button>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('Loading', () => {
  it('renders', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <Loading />, div)
  })
  test('snapshots', () => {
    const component = renderer.create(
      <Loading />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
describe('ButtonWithLoading', () => {
  it('renders', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <ButtonWithLoading />, div)
  })
  test('snapshots', () => {
    const component = renderer.create(
      <ButtonWithLoading />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' }
    ]
  }
  it('shows two items in list', () => {
    const element = shallow(
      <Table {...props} />
    )
    expect(element.find('.table-row').length).toBe(2)
  })
  it('renders', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Table {...props} />, div)
  })
  test('snapshots', () => {
    const component = renderer.create(
      <Table {...props} />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
