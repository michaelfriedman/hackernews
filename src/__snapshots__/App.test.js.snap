exports[`Button snapshots 1`] = `
<button
  className=""
  onClick={undefined}
  type="button">
  More
</button>
`;

exports[`ButtonWithLoading snapshots 1`] = `
<button
  className=""
  onClick={undefined}
  type="button" />
`;

exports[`Loading snapshots 1`] = `
<div>
  <span
    aria-hidden={true}
    className="fa fa-spinner fa-5x fa-spin fa fa-spinner"
    style={
      Object {
        "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
      }
    } />
</div>
`;

exports[`Search snapshots 1`] = `
<form
  onSubmit={undefined}>
  <input
    onChange={undefined}
    type="text"
    value={undefined} />
  <button
    type="submit">
    Search
  </button>
</form>
`;

exports[`Table snapshots 1`] = `
<div
  className="table">
  <div
    className="table-row">
    <span
      style={
        Object {
          "width": "40%",
        }
      }>
      <a
        href={undefined}>
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row">
    <span
      style={
        Object {
          "width": "40%",
        }
      }>
      <a
        href={undefined}>
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }>
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
</div>
`;

exports[`test snapshots 1`] = `
<div
  className="page">
  <div
    className="interactions">
    <form
      onSubmit={[Function]}>
      <input
        onChange={[Function]}
        type="text"
        value="redux" />
      <button
        type="submit">
        Search
      </button>
    </form>
  </div>
  <div
    className="table" />
  <div
    className="interactions">
    <div>
      <span
        aria-hidden={true}
        className="fa fa-spinner fa-5x fa-spin fa fa-spinner"
        style={
          Object {
            "textShadow": "0 1px 0 rgba(0, 0, 0, 0.1)",
          }
        } />
    </div>
  </div>
</div>
`;
